File: packages/sdk-flutter/rovenue_flutter_ios/ios/Classes/PaywallPlatformView.swift
// PaywallPlatformView.swift — hosts the SwiftUI `RovenuePaywallView`
// (packages/sdk-swift/Sources/Rovenue/PaywallUI/) inside a Flutter
// `FlutterPlatformView`. Task 7.
//
// Model: `packages/sdk-rn/ios/RovenuePaywallExpoView.swift`. Flutter's
// platform-view contract differs from Expo's in one load-bearing way that
// simplifies this file relative to that model: creation params arrive
// ONCE, atomically, at `init` — there is no per-prop update channel a
// Flutter host can use to change `placementIdentifier`/`locale`/etc. after
// the view exists (unlike Expo's `onViewDidUpdateProps`, which fires once
// per prop batch and can fire many times over a view's life). So there is
// no `reload()`/prop-diffing/content-key-cache dance here: the paywall is
// resolved exactly once, the result is cancelled on `deinit` if still in
// flight, and `mount` either builds the hosting controller (first
// successful resolve) or swaps its `rootView` (a mount from a background
// queue racing a cancelled task, defensively — in practice this only ever
// runs once per view).
//
// iOS-only — see `PaywallViewFactory.swift`'s header for why this file is
// excluded from the macOS SwiftPM test harness rather than branched with
// `#if os(macOS)`.

import Foundation
import UIKit
import SwiftUI
import Flutter
import Rovenue

/// Code sent for an `onPurchaseFailed` error that reached `fail(_:)`'s
/// generic (non-`RovenueError`) branch. `"Internal"` there is a REAL UDL
/// `ErrorKind` variant, so unlike the RN bridge's own sentinel this needs
/// no special casing here — `errorArgs(_:)` already produces the right
/// `code` for every error shape via `fail(_:)`.
final class PaywallPlatformView: NSObject, FlutterPlatformView {
  private let containerView = UIView()
  private let channel: FlutterMethodChannel

  private let placementIdentifier: String
  private let locale: String?
  private let colorSchemeOverride: String?
  private let hasRestoreHandler: Bool
  private let hasUrlHandler: Bool

  private var hostingController: UIHostingController<AnyView>?
  private var loadTask: Task<Void, Never>?

  init(frame: CGRect, viewId: Int64, args: Any?, messenger: FlutterBinaryMessenger) {
    let params = args as? [String: Any] ?? [:]
    self.placementIdentifier = params["placementIdentifier"] as? String ?? ""
    self.locale = params["locale"] as? String
    self.colorSchemeOverride = params["colorSchemeOverride"] as? String
    self.hasRestoreHandler = params["hasRestoreHandler"] as? Bool ?? false
    self.hasUrlHandler = params["hasUrlHandler"] as? Bool ?? false
    self.channel = FlutterMethodChannel(
      name: "dev.rovenue.flutter/paywall_view_\(viewId)",
      binaryMessenger: messenger
    )
    containerView.frame = frame
    containerView.clipsToBounds = true
    super.init()
    load()
  }

  func view() -> UIView { containerView }

  deinit {
    loadTask?.cancel()
  }

  private func load() {
    guard !placementIdentifier.isEmpty else { return }
    let placement = placementIdentifier
    let requestedLocale = locale
    loadTask = Task { [weak self] in
      // A resolution failure renders nothing — matches the RN bridge and
      // the SwiftUI renderer's own behavior for a nil/undecoded config.
      // There is no props contract here to preserve by staying silent
      // (unlike the RN bridge's comment on this point), it's simply that
      // this view has no error slot to surface one through.
      let paywall = try? await Rovenue.shared.getPaywall(placementId: placement, locale: requestedLocale)
      guard !Task.isCancelled else { return }
      await MainActor.run {
        self?.mount(paywall)
      }
    }
  }

  private func scheme() -> ColorScheme? {
    switch colorSchemeOverride {
    case "light": return .light
    case "dark": return .dark
    default: return nil
    }
  }

  private func mount(_ paywall: Paywall?) {
    guard let paywall else { return }

    let content = RovenuePaywallView(
      paywall: paywall,
      locale: locale,
      colorSchemeOverride: scheme(),
      onPurchaseCompleted: { [weak self] result in
        self?.channel.invokeMethod("onPurchaseCompleted", arguments: ["result": dtoFromPurchaseResult(result)])
      },
      onPurchaseFailed: { [weak self] error in
        self?.channel.invokeMethod("onPurchaseFailed", arguments: errorArgs(error))
      },
      onClose: { [weak self] in self?.channel.invokeMethod("onCloseRequested", arguments: nil) },
      onRestore: hasRestoreHandler ? { [weak self] in self?.channel.invokeMethod("onRestoreRequested", arguments: nil) } : nil,
      onUrl: hasUrlHandler ? { [weak self] url in
        self?.channel.invokeMethod("onUrlRequested", arguments: ["url": url.absoluteString])
      } : nil
    )

    if let hostingController {
      hostingController.rootView = AnyView(content)
      return
    }

    let controller = UIHostingController(rootView: AnyView(content))
    controller.view.backgroundColor = .clear
    controller.view.frame = containerView.bounds
    controller.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    containerView.addSubview(controller.view)
    hostingController = controller
  }
}

File: packages/sdk-flutter/rovenue_flutter_android/android/src/main/kotlin/dev/rovenue/flutter/PaywallPlatformView.kt
// PaywallPlatformView.kt — hosts the Android Views paywall renderer
// (dev.rovenue.sdk.paywallui.RovenuePaywallView) inside a Flutter
// `PlatformView`. Task 7.
//
// Model: `packages/sdk-rn/android/.../RovenuePaywallExpoView.kt`. Creation
// params arrive once, atomically, in the constructor — but unlike Expo's
// `onViewDidUpdateProps`, Flutter's `AndroidView`/`UiKitView` never resend
// them on a prop change. Task 8's carry-forward closes that gap with an
// explicit `updateParams` call the Dart side (`RovenuePaywallView`'s
// `didUpdateWidget`, see `packages/sdk-flutter/rovenue_flutter/lib/src/
// paywall_view.dart`) sends over this view's own per-instance channel —
// the same pattern `google_maps_flutter`/`webview_flutter` use for
// prop-diffing platform views. `onMethodCall` only re-resolves the paywall
// (cancelling any in-flight `loadJob`) when
// `placementIdentifier`/`locale`/`colorSchemeOverride` actually changed; a
// pure `hasRestoreHandler`/`hasUrlHandler` flip just re-binds the cached
// `currentPaywall` with new closures, so an unchanged prop never triggers
// a re-fetch.

package dev.rovenue.flutter

import android.content.Context
import android.view.View
import dev.rovenue.sdk.Rovenue
import dev.rovenue.sdk.paywallui.PaywallViewOptions
import dev.rovenue.sdk.paywallui.RovenuePaywallView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import dev.rovenue.sdk.Paywall

internal class PaywallPlatformView(
    context: Context,
    viewId: Int,
    args: Any?,
    messenger: BinaryMessenger,
) : PlatformView {

    private val inner = RovenuePaywallView(context)
    private val channel = MethodChannel(messenger, "dev.rovenue.flutter/paywall_view_$viewId")
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var loadJob: Job? = null

    private var placementIdentifier: String
    private var locale: String?
    private var colorSchemeOverride: String?
    private var hasRestoreHandler: Boolean
    private var hasUrlHandler: Boolean

    /** Cached so a handler-only `updateParams` (no placement/locale/scheme
     *  change) can re-`bind` without re-resolving the paywall. */
    private var currentPaywall: Paywall? = null

    init {
        @Suppress("UNCHECKED_CAST")
        val params = args as? Map<String, Any?> ?: emptyMap()
        placementIdentifier = params["placementIdentifier"] as? String ?: ""
        locale = params["locale"] as? String
        colorSchemeOverride = params["colorSchemeOverride"] as? String
        hasRestoreHandler = params["hasRestoreHandler"] as? Boolean ?: false
        hasUrlHandler = params["hasUrlHandler"] as? Boolean ?: false
        channel.setMethodCallHandler(::onMethodCall)
        load()
    }

    override fun getView(): View = inner

    override fun dispose() {
        loadJob?.cancel()
        scope.cancel()
    }

    /** Handles Dart's `updateParams` call (see this file's header). Any
     *  other method name is answered "not implemented" — this channel is
     *  also used the other way (native -> Dart `invokeMethod` calls in
     *  [options]), so an unrecognized method here isn't necessarily a bug.
     */
    private fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "updateParams") {
            result.notImplemented()
            return
        }
        @Suppress("UNCHECKED_CAST")
        val params = call.arguments as? Map<String, Any?> ?: emptyMap()
        val newPlacement = params["placementIdentifier"] as? String ?: ""
        val newLocale = params["locale"] as? String
        val newColorScheme = params["colorSchemeOverride"] as? String
        val newHasRestore = params["hasRestoreHandler"] as? Boolean ?: false
        val newHasUrl = params["hasUrlHandler"] as? Boolean ?: false

        val needsReload = newPlacement != placementIdentifier ||
            newLocale != locale ||
            newColorScheme != colorSchemeOverride
        val handlersChanged = newHasRestore != hasRestoreHandler || newHasUrl != hasUrlHandler

        placementIdentifier = newPlacement
        locale = newLocale
        colorSchemeOverride = newColorScheme
        hasRestoreHandler = newHasRestore
        hasUrlHandler = newHasUrl

        if (needsReload) {
            loadJob?.cancel()
            load()
        } else if (handlersChanged) {
            currentPaywall?.let { inner.bind(it, options()) }
        }
        result.success(null)
    }

    private fun load() {
        val placement = placementIdentifier
        if (placement.isEmpty()) return
        loadJob = scope.launch {
            // A resolution failure renders nothing — matches the RN bridge
            // and the Android renderer's own behavior for a null config.
            val paywall = runCatching { Rovenue.shared.getPaywall(placement, locale) }.getOrNull()
            if (paywall != null) {
                currentPaywall = paywall
                inner.bind(paywall, options())
            }
        }
    }

    private fun options(): PaywallViewOptions = PaywallViewOptions(
        locale = locale,
        darkMode = when (colorSchemeOverride) {
            "light" -> false
            "dark" -> true
            else -> null
        },
        onPurchaseCompleted = { result ->
            channel.invokeMethod("onPurchaseCompleted", mapOf("result" to dtoFromPurchaseResult(result)))
        },
        onPurchaseFailed = { error ->
            channel.invokeMethod("onPurchaseFailed", errorArgs(error))
        },
        onClose = { channel.invokeMethod("onCloseRequested", null) },
        // A null handler is what makes the native view HIDE restore
        // affordances / leave url buttons inert — so absence must survive
        // to `PaywallViewOptions`, not become a no-op lambda.
        onRestore = if (hasRestoreHandler) ({ channel.invokeMethod("onRestoreRequested", null) }) else null,
        onUrl = if (hasUrlHandler) ({ url -> channel.invokeMethod("onUrlRequested", mapOf("url" to url)) }) else null,
    )
}
